src/trs/hmtrs.rs

```diff
@@ -109,11 +167,8 @@ impl HMTRS {
             Term::Application { op, args } => {
                 let body_type = self.infer_args(args, ctx)?;
                 let head_type = self.infer_op(op)?.instantiate(ctx).apply(ctx);
-                if ctx.unify(&head_type, &body_type).is_ok() {
-                    Ok(self.infer_op(op)?.instantiate(ctx).apply(ctx))
-                } else {
-                    Err(TypeError)
-                }
+                ctx.unify(&head_type, &body_type)?;
+                Ok(self.infer_op(op)?.instantiate(ctx).apply(ctx))
```

Comment:
There's a function somewhere in here called `fast_update` that manages this series of `infer_*`, `instantiate`, and `apply`. The name isn't a great name, but it might be good to use that (renamed) function consistently to provide a bit less verbose and more semantic description of these inference algorithms.